### ISSUE TITLE
UPX compressed binaries don't work with qemu-i386

### ISSUE THREAD

**Issue by markus-oberhumer** [upx/upx#685]
UPDATE: this was a UPX bug that was fixed in UPX v3.99; physical machines and older qemu versions tend to return mmap() ranges to hide the problem, but it manifested with newer qemu versions

## What's the problem (or question)?

UPX compressed binaries don't work with qemu-i386 version 5.2.0 or later. 

qemu-i386 4.2.0 and 2.12.0 work, so this is a possible regression.

Tested with qemu-2.12.0 (Alpine Linux 3.8), qemu-3.1.0 (Alpine Linux 3.9), qemu-4.2.0 (Alpine Linux 3.11) and lots of other qemu versions including current git HEAD and `upx-fix` branch.

Affected files are upx i386_linux 3.95 and 3.96:
- https://github.com/upx/upx/releases/download/v3.95/upx-3.95-i386_linux.tar.xz
- https://github.com/upx/upx/releases/download/v3.96/upx-3.96-i386_linux.tar.xz

(Possibly related, but highly unlikely because this is qemu-ppc):
- https://github.com/upx/upx/releases/download/v3.94/upx-3.94-powerpc_linux.tar.xz

## How can we reproduce the issue?

```sh
$ ./upx-3.95-i386_linux/upx
# [WORKS]
$ qemu-i386 ./upx-3.95-i386_linux/upx
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
zsh: segmentation fault (core dumped)  qemu-i386 ./upx-3.95-i386_linux/upx

$ ./upx-3.96-i386_linux/upx
# [WORKS]
$ qemu-i386 ./upx-3.96-i386_linux/upx
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
zsh: segmentation fault (core dumped)  qemu-i386 ./upx-3.96-i386_linux/upx

```

Hint: see https://github.com/upx/upx/blob/devel/misc/test-qemu-with-podman/README.md for some support scripts how to test QEMU under Podman

P.S. I stumbled on this while testing the fix for #683

**Comment by markus-oberhumer** [comment]
CC @hdeller Helge please have a look if you're interested. Thanks!

**Comment by hdeller** [comment]
I see that upx -3.95 does a check on the value of the first mmap returned, and then hlts if it's below 0xfffff000.
I think this is actually wrong behavior from upx, since an anonymous mapping to (NULL) can return any address and upx should not depend on that it's some specific value. 
Maybe I can workaround that in qemu...?

### 139961 mmap(NULL,442340,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANONYMOUS,-1,0) = 0x40804000
----------------
IN:
0x080b3923:  3d 00 f0 ff ff           cmpl /local/s$0xfffff000, %eax
0x080b3928:  7c 01                    jl /local/s0x80b392b
----------------
IN:  0x080b392a:  f4                       hlt /local/s
--- SIGSEGV {si_signo=SIGSEGV, si_code=SI_KERNEL, si_pid=0, si_uid=0} ---

**Comment by hdeller** [comment]
This qemu hack makes it run: (copy and pasted here)

diff --git a/linux-user/syscall.c b/linux-user/syscall.c
index aa906bedcc..bb65a1bd1d 100644
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -10576,6 +10577,7 @@ static abi_long do_syscall1(CPUArchState *cpu_env, int num, abi_long arg1,
             v5 = tswapal(v[4]);
             v6 = tswapal(v[5]);
             unlock_user(v, arg1, 0);
+if (!v1) v1=0x0f0000000;
             ret = get_errno(target_mmap(v1, v2, v3,
                                         target_to_host_bitmask(v4, mmap_flags_tbl),
                                         v5, v6));

**Comment by hdeller** [comment]
Adding @jreiser, what does this i386-assembly means? (I know x86 assembly, but don't do it on daily basis)

0x080b3923:  3d 00 f0 ff ff           cmpl $0xfffff000, %eax
0x080b3928:  7c 01                    jl 0x80b392b
0x080b392a:  f4                       hlt 

%eax is the return value from "mmap(NULL, 442340, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0)".
with %eax = 0x40804000,  qemu will stop at the "hlt".
with %eax = 0xf7f4a000, qemu will continue at 0x80b392b.

IMHO, both values are less than 0xfffff000, so in both cases qemu should continue at 0x80b392b.
Is this maybe a qemu bug, where the translator generates wrong code for cmp/jl with 32-bit 386 assembly ?

**Comment by hdeller** [comment]
this is what qemu generates:
IN: 
0x080b3923:  3d 00 f0 ff ff           cmpl $0xfffff000, %eax
0x080b3928:  7c 01                    jl 0x80b392b

OUT: [size=115]
  -- guest addr 0x00000000080b3923 + tb prologue
0x7f8bdc000980:  8b 5d f0                 movl -0x10(%rbp), %ebx
0x7f8bdc000983:  85 db                    testl %ebx, %ebx
0x7f8bdc000985:  0f 8c 5c 00 00 00        jl 0x7f8bdc0009e7
0x7f8bdc00098b:  c7 45 2c 00 f0 ff ff     movl $0xfffff000, 0x2c(%rbp)
0x7f8bdc000992:  8b 5d 00                 movl (%rbp), %ebx
0x7f8bdc000995:  44 8b e3                 movl %ebx, %r12d
0x7f8bdc000998:  41 81 ec 00 f0 ff ff     subl $0xfffff000, %r12d
0x7f8bdc00099f:  44 89 65 28              movl %r12d, 0x28(%rbp)
  -- guest addr 0x00000000080b3928
0x7f8bdc0009a3:  c7 45 34 10 00 00 00     movl $0x10, 0x34(%rbp)
0x7f8bdc0009aa:  81 fb 00 f0 ff ff        cmpl $0xfffff000, %ebx
0x7f8bdc0009b0:  0f 8c 19 00 00 00        jl 0x7f8bdc0009cf
0x7f8bdc0009b6:  90                       nop 
0x7f8bdc0009b7:  e9 00 00 00 00           jmp 0x7f8bdc0009bc
0x7f8bdc0009bc:  c7 45 20 2a 39 0b 08     movl $0x80b392a, 0x20(%rbp)
0x7f8bdc0009c3:  48 8d 05 f7 fe ff ff     leaq -0x109(%rip), %rax
0x7f8bdc0009ca:  e9 49 f6 ff ff           jmp 0x7f8bdc000018
0x7f8bdc0009cf:  e9 00 00 00 00           jmp 0x7f8bdc0009d4
0x7f8bdc0009d4:  c7 45 20 2b 39 0b 08     movl $0x80b392b, 0x20(%rbp)
0x7f8bdc0009db:  48 8d 05 de fe ff ff     leaq -0x122(%rip), %rax
0x7f8bdc0009e2:  e9 31 f6 ff ff           jmp 0x7f8bdc000018
0x7f8bdc0009e7:  48 8d 05 d5 fe ff ff     leaq -0x12b(%rip), %rax
0x7f8bdc0009ee:  e9 25 f6 ff ff           jmp 0x7f8bdc000018

**Comment by markus-oberhumer** [comment]
from upx-3.95-src/src/stub/src/i386-linux.elf-entry.S at line 252:

```asm

mmap:  // oldmmap: ebx -> 6 arguments; remove arguments on return
        lea ebx,[4+esp]
        push __NR_mmap; pop eax; int 0x80
        cmp eax,PAGE_MASK; jl 0f; hlt; 0:
        ret 6*4

```

**Comment by markus-oberhumer** [comment]
Update: this is probably a UPX bug, fixed in 3.99:

See commit https://github.com/upx/upx/commit/7626e09f665b765e0328ef851e2bb06604bf9edd

(this changes the compare from signed to unsigned)

**Comment by markus-oberhumer** [comment]
@jreiser please confirm this is an UPX bug that was unnoticed for a very long time by pure mmap() accident

**Comment by jreiser** [comment]
Yes. this **was** a UPX bug.  i386 "jl" is signed; was fixed in https://github.com/upx/upx/commit/7626e09f665b765e0328ef851e2bb06604bf9edd to be "jb" unsigned.  (typo/thinko/too_tired)

**Comment by markus-oberhumer** [comment]
@hdeller Helge, many thanks for looking into this!

Closing as a bug that was fixed in UPX v3.99.
